feat: Support disjunctive authorization checks in the Engine

ce-dmelnych opened the pull request:
## Description

This PR introduces support for disjunctive (OR) authorization checks in the Zeebe Engine, enabling scenarios where a user needs to be authorized through at least one of multiple possible authorization paths. 

### Changes

1. (Refactoring) Extracted `RejectionAggregator` - extracted rejection aggregation logic from `AuthorizationCheckBehavior` into a dedicated utility class with:
   - `aggregate()` - Combines rejections for single authorization checks (existing logic)
   - `aggregateComposite()` - Combines rejections from multiple authorization paths using `"; and "` separator

2. Added new authorization methods to `AuthorizationCheckBehavior`:
   - `isAnyAuthorized(AuthorizationRequest...)` - Returns success if any request is authorized
   - `isAnyAuthorizedOrInternalCommand(AuthorizationRequest...)` - Same as above but bypasses auth for internal commands

### Example usage

```java
// User can read a user task if they have EITHER: 
// - READ_USER_TASK permission on the process definition, OR
// - READ user task
authorizationCheckBehavior.isAnyAuthorized(
    AuthorizationRequest.builder()
        .command(command)
        .resourceType(AuthorizationResourceType.PROCESS_DEFINITION)
        .permissionType(PermissionType.READ_USER_TASK)
        .addResourceId(processDefinitionId)
        .build(),
    AuthorizationRequest.builder()
        .command(command)
        .resourceType(AuthorizationResourceType.USER_TASK)
        .permissionType(PermissionType.READ)
        .resourceProperties(...)
        .build()
);
```

## Related issues

closes #40112

ce-dmelnych committed:
refactor: extract RejectionAggregator from AuthorizationCheckBehavior

Extract rejection aggregation logic to dedicated component:
- aggregate(): combine rejections for single authorization check

Prioritizes permission rejections over tenant rejections.

ce-dmelnych committed:
feat: add aggregateComposite method to RejectionAggregator

Add aggregateComposite() method to support combining rejections from
multiple authorization paths (ANY/OR logic). When all paths fail,
the method combines messages with "; and " separator and provides
`FORBIDDEN` rejection type if any rejection is `FORBIDDEN`.

ce-dmelnych committed:
feat: add isAnyAuthorized methods for disjunctive authorization checks

Add `isAnyAuthorized` and `isAnyAuthorizedOrInternalCommand` methods to
`AuthorizationCheckBehavior` class to provide for support OR/disjunctive authorization logic where at least
one of multiple authorization requests must succeed.

- Add `isAnyAuthorized(AuthorizationRequest...)` for checking if any request passes
- Add `isAnyAuthorizedOrInternalCommand(AuthorizationRequest...)` to bypass auth for internal commands
- Use `RejectionAggregator.aggregateComposite()` to combine rejection messages with "; and " separator
- Add tests for empty requests, first/second request authorization, and combined rejection

Ben-Sheppard commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/aggregator/RejectionAggregator.java:
```diff
@@ -0,0 +1,96 @@
+    // Should not happen, but fallback to forbidden
+    return Either.left(
+        new Rejection(RejectionType.FORBIDDEN, "Authorization failed for unknown reason"));
+  }
```
🔧 If this shouldn't happen then I'd suggest we can remove it. In any case, the returned message, because of the nature, doesn't allow me to know how to fix the issue.

Ben-Sheppard reviewed (commented):
Just a few comments mostly on small code style points but otherwise the change looks fine :)

ce-dmelnych committed:
refactor: fail fast on invalid rejection aggregation inputs

Replace silent fallback rejections with exceptions for invalid states:
- Throw `IllegalArgumentException` for empty rejection lists
- Throw `IllegalStateException` when rejection is neither permission nor tenant

These cases indicate programming errors and should fail loudly rather than
returning generic `FORBIDDEN` rejections that obscure the root cause.

Also extract common filtering logic into `aggregateMatchingRejections()`
and simplify return type from `Either<Rejection, Void>` to `Rejection`.

ce-dmelnych committed:
refactor: delegate isAuthorizedOrInternalCommand to isAnyAuthorizedOrInternalCommand

Consolidate authorization logic by having the single-request method
delegate to the varargs method, ensuring consistent behavior and
reducing code duplication.

ce-dmelnych committed:
refactor: throw exception for empty authorization requests

Replace FORBIDDEN rejection with IllegalArgumentException when no
authorization requests are provided to isAnyAuthorized() and
isAnyAuthorizedOrInternalCommand().

Empty request arrays indicate a programming error, not a user
authorization failure, and should fail fast rather than return
a misleading rejection.

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/aggregator/RejectionAggregator.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    // Fallback: return the first rejection if present
+    if (!rejections.isEmpty()) {
```
Currently we could only have permission and tenant rejections. I've removed this fallback.

ce-dmelnych commented:
> Just a few comments mostly on small code style points but otherwise the change looks fine :)

Thanks for the review, @Ben-Sheppard! I've addressed all your comments and applied the suggestions.
Would you mind taking a look at the last 3 commits when you get a chance? 🙏

Ben-Sheppard reviewed (approved):
🚀 Thank you for addressing my comments, your changes commits look great to me. 

💭 I had wondered if there was a strong need to separate out internal commands or if we can just include that check in one single method but I don't know enough about if we would have internal commands that need authorization checks. I don't see this as a blocker though.
